File: src/main/java/com/merricklabs/portlandbus/external/trimet/models/Arrival.java
package com.merricklabs.portlandbus.external.trimet.models;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.Arrays;
import java.util.Date;
import lombok.extern.slf4j.Slf4j;

import static java.time.temporal.ChronoUnit.MINUTES;

@Slf4j
@JsonIgnoreProperties(ignoreUnknown = true)
public class Arrival {

    private Date scheduled;
    private Date estimated;
    private String shortSign;
    private int locid;
    private LocalDateTime time;

    public int getStopId(){
        return locid;
    }

    public long getMinutesRemaining(LocalDateTime now){
        return now.until(this.getTime(), MINUTES);
    }

    public int getBusId(){
        // This field looks like "20 to Gresham"
        return Integer.parseInt(Arrays.asList(shortSign.split(" ")).get(0));
    }

    /**
     * Returns the estimated time, if present. Otherwise, returns the scheduled time.
     * Essentially, this returns the best guess at the arrival time.
     * @return
     */
    public LocalDateTime getTime(){
        // Lazily init time
        if (time == null) {
            time = estimated != null ? convertDate(estimated) : convertDate(scheduled);
            log.info("Set time of this arrival to {}", time);
        }

        return time;
    }

    private LocalDateTime convertDate(Date date){
        return date.toInstant().atZone(ZoneId.systemDefault()).toLocalDateTime();
    }

    public Arrival(
            @JsonProperty("scheduled")
                    Date scheduled,
            @JsonProperty("estimated")
                    Date estimated,
            @JsonProperty("locid")
                    int locid,
            @JsonProperty("shortSign")
                    String shortSign
    ) {
        this.scheduled = scheduled;
        this.estimated = estimated;
        this.locid = locid;
        this.shortSign = shortSign;
    }
}

File: src/main/kotlin/com/merricklabs/portlandbus/external/trimet/models/Arrival.kt
package com.merricklabs.portlandbus.external.trimet.models

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.temporal.ChronoUnit.MINUTES
import java.util.*

@JsonIgnoreProperties(ignoreUnknown = true)
class Arrival(
        @param:JsonProperty("scheduled")
        private val scheduled: Date,
        @param:JsonProperty("estimated")
        private val estimated: Date?,
        @param:JsonProperty("locid")
        val stopId: Int,
        @param:JsonProperty("shortSign")
        private val shortSign: String
) {

    /**
     * Returns the estimated time, if present. Otherwise, returns the scheduled time.
     * Essentially, this returns the best guess at the arrival time.
     * @return
     */
    private var _time: LocalDateTime? = null
    val time: LocalDateTime
        get() {
            // Lazily init time
            if (_time == null) {
                _time = if (estimated != null) convertDate(estimated) else convertDate(scheduled)
                // log.debug("Set time of this arrival to {}", time)
            }
            return _time ?: throw RuntimeException("Time should not be null.")
        }

    // This field looks like "20 to Gresham"
    val busId: Int
        get() =
            Integer.parseInt(Arrays.asList(*shortSign.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray())[0])

    fun getMinutesRemaining(now: LocalDateTime): Long {
        return now.until(this.time, MINUTES)
    }

    private fun convertDate(date: Date): LocalDateTime {
        return date.toInstant().atZone(ZoneId.systemDefault()).toLocalDateTime()
    }
}
